### PR TITLE
Python: Implement AI request settings

### PR DESCRIPTION
**Pull request by eavanvalkenburg**
### Motivation and Context

This PR implements the new approach to AI Request settings as described here: https://github.com/microsoft/semantic-kernel/blob/main/docs/decisions/0008-support-generic-llm-request-settings.md and completes #3312 

### Description

Does this:
- Creates AI Request Settings with service_id (referring to the registered AI services in a kernel) and extension_data, a dict that can hold any number of settings
- has several methods:
  - `from_ai_request_settings` -> constructor to create a new request_settings object based on another, supports creating specific ai_request_settings from generic ones.
  - `update_from_ai_request_settings` -> updating the fields of a request setting from another ones.
  - `prepare_settings_dict` -> method that is used when actually creating the dict that get's passed to the handler of the AI service.
- Created subclasses like OpenAIChatRequestSettings which has all the required fields of a OpenAI Chat completion, same for OpenAIText, AzureChat, GooglePalm, HuggingFace, etc. 
- added `get_request_settings_class` to ai_services to return the class used for the settings for that service.

The goal is that when running, this settings object get's updated and a single operations (called `prepare_settings_dict`) is used  to return a dict that can be passed to the api call directly, thereby reducing the need for custom logic everywhere, this will also negate the need for things like `complete_chat_with_functions_async` and others. So also removed complete_chat_..._async methods from openai and azure openai.

It has also changed the prompt_template_config with the generic type for the AIRequestSettings or subclasses, and some other changes.

### Contribution Checklist

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [SK Contribution Guidelines](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/semantic-kernel/blob/main/CONTRIBUTING.md#development-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [ ] I didn't break anyone :smile:

### CODE REVIEW

File: python/semantic_kernel/connectors/ai/ai_request_settings.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
+from typing import Any, Dict, Optional
+
+from pydantic import Field
+
+from semantic_kernel.sk_pydantic import SKBaseModel
+
+
+class AIRequestSettings(SKBaseModel):
+    service_id: Optional[str] = Field(None, min_length=1)
```
Comment (moonbox3) [inline python/semantic_kernel/connectors/ai/ai_request_settings.py:9]: Could we add some comments here about what the service_id and extension_data is, how they're used, etc?

File: python/semantic_kernel/connectors/ai/google_palm/services/gp_text_completion.py [hunk 2]
```diff
@@ -86,24 +50,27 @@ async def _send_completion_request(
...
-                stop_sequences=(
-                    request_settings.stop_sequences
-                    if request_settings.stop_sequences is not None
-                    and len(request_settings.stop_sequences) > 0
-                    else None
-                ),
-                candidate_count=request_settings.number_of_responses,
-                top_p=request_settings.top_p,
-            )
+            response = palm.generate_text(**request_settings.prepare_settings_dict())
```
Comment (moonbox3) [inline python/semantic_kernel/connectors/ai/google_palm/services/gp_text_completion.py:53]: Love the de-cluttering here. Amazing.

**Comment by moonbox3** [comment]
What an awesome PR. Thank you for doing this. I love the direction we're heading and how much you're simplifying config related items.

**Comment by moonbox3** [comment]
Have we gone through all Kernel examples and updated them with these new settings? Does the example `chat_gpt_api_function_calling` need any update regarding the prompt template config.

File: python/samples/kernel-syntax-examples/azure_chat_gpt_api.py [hunk 3]
```diff
@@ -25,11 +26,14 @@
     **azure_openai_settings_from_dot_env_as_dict(include_api_version=True)
 )
 kernel.add_chat_service("chat-gpt", chat_service)
-
-prompt_config = sk.PromptTemplateConfig.from_completion_parameters(
-    max_tokens=2000, temperature=0.7, top_p=0.8
+req_settings = AIRequestSettings(
```
[the remark this answers is not in the document]
Comment (eavanvalkenburg) [reply python/samples/kernel-syntax-examples/azure_chat_gpt_api.py:29]: I need to add some notes on the logic here, but it is also possible to get the request settings class from the service, which is a nice way of doing it when just using one service, the advantage of the whole approach is that you can define the parameters for multiple types of services and they can parsed and matched to the specific settings dynamically during runtime.

File: python/semantic_kernel/connectors/ai/open_ai/services/open_ai_text_completion_base.py [hunk 4]
```diff
@@ -57,15 +65,15 @@ async def complete_stream_async(
...
+            settings {OpenAIRequestSettings} -- The settings to use for the completion request.
             logger {Optional[Logger]} -- The logger instance to use. (Optional)
 
         Returns:
             Union[str, List[str]] -- The completion result(s).
         """
-        response = await self._send_request(
-            prompt=prompt, request_settings=settings, stream=True
-        )
+        settings.prompt = prompt
```
Comment (moonbox3) [inline python/semantic_kernel/connectors/ai/open_ai/services/open_ai_text_completion_base.py:74]: The incoming settings of type `OpenAIRequestSettings` doesn't look to be of an OpenAITextRequestSettings instance, so it doesn't have a prompt attribute to set.

Comment (moonbox3) [reply python/semantic_kernel/connectors/ai/open_ai/services/open_ai_text_completion_base.py:74]: Even with that update, I see this error:

TypeError("Missing required arguments; Expected either ('messages' and 'model') or ('messages', 'model' and 'stream') arguments to be given")

File: python/samples/kernel-syntax-examples/azure_chat_gpt_api.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (eavanvalkenburg) [reply python/samples/kernel-syntax-examples/azure_chat_gpt_api.py]: So I added a bit of text to that sample with the different ways you can setup the request settings and when to choose which.

File: python/semantic_kernel/connectors/ai/open_ai/request_settings/azure_chat_request_settings.py [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+from pydantic import Field
+from pydantic.dataclasses import dataclass
+
+from semantic_kernel.connectors.ai.open_ai.request_settings.open_ai_request_settings import (
+    OpenAIChatRequestSettings,
+)
+from semantic_kernel.sk_pydantic import SKBaseModel
+
+logger = logging.getLogger(__name__)
```
Comment (juliomenendez) [inline python/semantic_kernel/connectors/ai/open_ai/request_settings/azure_chat_request_settings.py:12]: If this is not being used in this module, do you need it?

File: python/semantic_kernel/connectors/ai/open_ai/semantic_functions/open_ai_chat_prompt_template.py [hunk 6]
```diff
@@ -43,19 +42,26 @@ def add_message(
...
+                        role=role,
+                        fixed_content=message,
+                        name=name,
+                        function_call=function_call,
+                    )
                 )
-                function_call = None
-        self._messages.append(
+                return
+            self._log.warning(
```
Comment (juliomenendez) [inline python/semantic_kernel/connectors/ai/open_ai/semantic_functions/open_ai_chat_prompt_template.py:56]: This should be `logger.warning`

File: python/semantic_kernel/connectors/ai/open_ai/semantic_functions/open_ai_chat_prompt_template_with_data_config.py [hunk 7]
```diff
@@ -1,106 +1,99 @@
 # Copyright (c) Microsoft. All rights reserved.
```
Comment (juliomenendez) [inline python/semantic_kernel/connectors/ai/open_ai/semantic_functions/open_ai_chat_prompt_template_with_data_config.py:1]: Given that the entire file is commented out, should it just be deleted?

File: python/tests/unit/ai/google_palm/services/test_palm_chat_completion.py [hunk 8]
```diff
@@ -65,17 +64,16 @@ async def test_google_palm_text_completion_complete_chat_async_call_with_paramet
             ai_model_id=ai_model_id,
             api_key=api_key,
         )
-        settings = ChatRequestSettings()
+        settings = GooglePalmChatRequestSettings()
         response = await gp_chat_completion.complete_chat_async(prompt, settings)
         assert isinstance(response.result(), str) and len(response.result()) > 0
-
+        print(mock_gp.chat)
```
Comment (juliomenendez) [inline python/tests/unit/ai/google_palm/services/test_palm_chat_completion.py:70]: Should this `print()` statement be here?

**Review by juliomenendez — CHANGES_REQUESTED** [review]
Looking good, just a couple of questions

File: python/semantic_kernel/connectors/ai/google_palm/services/gp_text_embedding.py [hunk 9]
```diff
@@ -12,6 +14,11 @@
 from semantic_kernel.connectors.ai.embeddings.embedding_generator_base import (
     EmbeddingGeneratorBase,
 )
+from semantic_kernel.connectors.ai.google_palm.request_settings.gp_request_settings import (
+    GooglePalmEmbeddingRequestSettings,
+)
+
+logger = logging.getLogger(__name__)
```
Comment (juliomenendez) [inline python/semantic_kernel/connectors/ai/google_palm/services/gp_text_embedding.py:21]: Do you need this if it's not used in this module?

File: python/semantic_kernel/connectors/ai/hugging_face/__init__.py [hunk 10]
```diff
@@ -1,6 +1,6 @@
 # Copyright (c) Microsoft. All rights reserved.
 
-from semantic_kernel.connectors.ai.hugging_face.hf_request_settings import (
+from connectors.ai.hugging_face.request_settings.hf_request_settings import (
```
Comment (juliomenendez) [inline python/semantic_kernel/connectors/ai/hugging_face/__init__.py:3]: All the other imports are using the absolute name with `semantic_kernel.`, shouldn't this follow the same?

File: python/semantic_kernel/connectors/ai/hugging_face/services/hf_text_completion.py [hunk 11]
```diff
@@ -9,7 +9,7 @@
 
 from semantic_kernel.connectors.ai.ai_exception import AIException
 from semantic_kernel.connectors.ai.ai_service_client_base import AIServiceClientBase
-from semantic_kernel.connectors.ai.hugging_face.hf_request_settings import (
```
Comment (juliomenendez) [inline python/semantic_kernel/connectors/ai/hugging_face/services/hf_text_completion.py]: All the other imports are using the absolute name with `semantic_kernel.`, shouldn't this follow the same?

**Review by juliomenendez — CHANGES_REQUESTED** [review]

File: python/semantic_kernel/connectors/ai/hugging_face/__init__.py (hunk 10 above)
Comment (eavanvalkenburg) [reply python/semantic_kernel/connectors/ai/hugging_face/__init__.py]: yeah, that is a mistake

**Review by juliomenendez — APPROVED** [review]
Looks good!

**Review by alexchaomander — APPROVED** [review]

**Review by moonbox3 — APPROVED** [review]

**Pull request merged**